src/Admin/AdminHelper.php

```diff
@@ -98,19 +98,36 @@ public function getAdmin($code)
      * Note:
      *   This code is ugly, but there is no better way of doing it.
      *
-     * @param object $subject
+     * @param object $deprecatedSubject
      * @param string $elementId
      *
      * @throws \RuntimeException
      * @throws \Exception
      *
      * @return array
      */
-    public function appendFormFieldElement(AdminInterface $admin, $subject, $elementId)
+    public function appendFormFieldElement(AdminInterface $admin, $deprecatedSubject, $elementId) // NEXT_MAJOR remove the $subject param
     {
         // child rows marked as toDelete
         $toDelete = [];
+
         // retrieve the subject
+        $subject = $admin->getSubject();
+        if (null === $subject) {
+            $subject = $deprecatedSubject;
+            @trigger_error(sprintf(
+                'Calling %s() when there is no suject set in the admin is deprecated since sonata-project/admin-bundle 3.x and will throw an exception in 4.0. '.
```

Comment:
```suggestion
                'Calling %s() when there is no subject set in the admin is deprecated since sonata-project/admin-bundle 3.x and will throw an exception in 4.0. '.
```